src/Compilers/CSharp/Portable/Binder/UsingStatementBinder.cs

```diff
@@ -149,6 +121,49 @@ internal override BoundStatement BindUsingStatementParts(DiagnosticBag diagnosti
                 boundBody,
                 awaitOpt,
                 hasErrors);
+
+            bool initConversion(bool fromExpression)
+            {
+                HashSet<DiagnosticInfo> useSiteDiagnostics = null;
+
+                iDisposableConversion = fromExpression ?
+                    originalBinder.Conversions.ClassifyImplicitConversionFromExpression(expressionOpt, disposableInterface, ref useSiteDiagnostics) :
+                    originalBinder.Conversions.ClassifyImplicitConversionFromType(declarationTypeOpt, disposableInterface, ref useSiteDiagnostics);
+
+                diagnostics.Add(fromExpression ? (CSharpSyntaxNode)expressionSyntax : declarationSyntax, useSiteDiagnostics);
+
+                if (iDisposableConversion.IsImplicit)
+                {
+                    return true;
+                }
+
+                TypeSymbol type = fromExpression ? expressionOpt.Type : declarationTypeOpt;
+                if (type is null || !type.IsErrorType())
+                {
+                    // Retry with a different assumption about whether the `using` is async
+                    TypeSymbol alternateInterface = getDisposableInterface(!hasAwait);
+                    HashSet<DiagnosticInfo> ignored = null;
+                    Conversion alternateConversion = fromExpression ?
+                        originalBinder.Conversions.ClassifyImplicitConversionFromExpression(expressionOpt, alternateInterface, ref ignored) :
+                        originalBinder.Conversions.ClassifyImplicitConversionFromType(declarationTypeOpt, alternateInterface, ref ignored);
+
+                    bool wrongAsync = alternateConversion.IsImplicit;
+                    ErrorCode errorCode = wrongAsync
+                        ? (hasAwait ? ErrorCode.ERR_NoConvToIAsyncDispWrongAsync : ErrorCode.ERR_NoConvToIDispWrongAsync)
+                        : (hasAwait ? ErrorCode.ERR_NoConvToIAsyncDisp : ErrorCode.ERR_NoConvToIDisp);
+
+                    Error(diagnostics, errorCode, (CSharpSyntaxNode)declarationSyntax ?? expressionSyntax, declarationTypeOpt ?? expressionOpt.Display);
+                }
+
+                return false;
```

Comment:
:memo: I've changed the bool value (flipping from true to false for error case, and using false for both error cases). It has no visible effect, but it's clearer.